### ISSUE TITLE
Learning to Continually Learn

### ISSUE THREAD

**Issue by juanmc2005** [juanmc2005/phd-bib#34]
## [Learning to Continually Learn](https://arxiv.org/abs/2002.09571)
*Continual lifelong learning requires an agent or model to learn many sequentially ordered tasks, building on previous knowledge without catastrophically forgetting it. Much work has gone towards preventing the default tendency of machine learning models to catastrophically forget, yet virtually all such work involves manually-designed solutions to the problem. We instead advocate meta-learning a solution to catastrophic forgetting, allowing AI to learn to continually learn. Inspired by neuromodulatory processes in the brain, we propose A Neuromodulated Meta-Learning Algorithm (ANML). It differentiates through a sequential learning process to meta-learn an activation-gating function that enables context-dependent selective activation within a deep neural network. Specifically, a neuromodulatory (NM) neural network gates the forward pass of another (otherwise normal) neural network called the prediction learning network (PLN). The NM network also thus indirectly controls selective plasticity (i.e. the backward pass of) the PLN. ANML enables continual learning without catastrophic forgetting at scale: it produces state-of-the-art continual learning performance, sequentially learning as many as 600 classes (over 9,000 SGD updates).*

### Cite this paper
```bibtex
@misc{beaulieu2020learning,
    title={Learning to Continually Learn},
    author={Shawn Beaulieu and Lapo Frati and Thomas Miconi and Joel Lehman and Kenneth O. Stanley and Jeff Clune and Nick Cheney},
    year={2020},
    eprint={2002.09571},
    archivePrefix={arXiv},
    primaryClass={cs.LG}
}
```

**Comment by juanmc2005** [comment]
### Description

This paper introduces a novel way to deal with catastrophic forgetting based on the OML method (Online Meta-Learning). Their experiences are performed on the Omniglot dataset in a sequential training setting where each class is seen only once. They find their model to perform better than all other baseline models, including a model trained sequentially from scratch, a pretrained model with transfer, the OML model itself and "oracles" trained in an i.i.d. fashion (to serve as upper bound).

What has not been tested with this approach is sequential learning across different domains, since the experiments are performed showing different classes from the same dataset sequentially.

### Architecture

The architecture they use consists of 2 parallel neural networks with 3 convolutional layers, a batch norm layer and a feed forward layer each.
The first network modulates the second one in the last layer using a simple element-wise multiplication.

![image](https://user-images.githubusercontent.com/5760780/78761417-d7b69d00-7982-11ea-8a35-16e3d6a7159c.png)

### Training

I won't describe the training algorithm here as it is quite complex and very well explained in the paper in section 3.2.

What's important to remember is that there's a meta-training and meta-testing phase. In the meta-training phase, 2 loops exist, the inner controlling the modulated model, and the outer the modulating model. During inner training, the outer network's weights are frozen, and only updated in the outer loop using the current training set and a remember set randomly sampled from the entire dataset (and not from the past only, which is not realistic in my view). In the meta-testing phase, the inner network is fine-tuned to other instances (from the same classes) unseen during training and then evaluates the whole network on the fine-tuning training set and a held out test set.

### Results

As the following figure shows, the ANML model outperforms all other baselines on the final test set, even their oracles (but not his own).
![image](https://user-images.githubusercontent.com/5760780/78764040-421d0c80-7986-11ea-8a95-cbddac8b8ae7.png)

Another very interesting result is that the application of the neuromodulatory signal results in sparsity that was not forced by any specific algorithm.
![image](https://user-images.githubusercontent.com/5760780/78764249-932d0080-7986-11ea-82f1-b967e1f3125f.png)

Finally, they observed that KNN classification accuracy was improved significantly after applying the neuromodulated signal, which means that the outer network forces activations to be more separable. This can be an interesting direction for representation learning as well.
![image](https://user-images.githubusercontent.com/5760780/78764562-05054a00-7987-11ea-9cbe-7388d7c7d376.png)

### Takeaways

The approach is really interesting, as existing approaches for continual learning are all based on hand-crafted algorithms that modify or constrain the network to fit a particular property hoping that it will lead to avoiding catastrophic forgetting.

It will be very interesting to try such an approach with speech or text, but with different datasets, in order to evaluate how well such a network can perform on different domains and which are its limits. It would also be interesting to evaluate on the first datasets of the experiments to verify that the network hasn't forgotten.

It is surprising how the neuromodulatory network acts as an inhibitor of the inner network, producing sparse activations in the last layers. This mean that the model wasn't optimized for sparsity, but it has found that itself with no dead neurons (that other sparsity oriented algorithms display).

Finally, another interesting thing to try is to modulate other (or more) layers in the inner network, maybe an embedding layer? This could be really fun.
